example/poisson/poisson_package.cpp

```diff
@@ -0,0 +1,156 @@
+template <typename T>
+TaskStatus UpdatePhi(T *u, T *du) {
+  Kokkos::Profiling::pushRegion("Task_Poisson_UpdatePhi");
+  auto pm = u->GetParentPointer();
+
+  IndexRange ib = u->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = u->GetBoundsK(IndexDomain::interior);
+
+  PackIndexMap imap;
+  std::vector<std::string> vars({"density", "potential"});
+  auto v = u->PackVariables(vars, imap);
+  const int irho = imap["density"].first;
+  const int iphi = imap["potential"].first;
+  std::vector<std::string> phi_var({"potential"});
+  auto dv = du->PackVariables(phi_var);
+
+  auto coords = GetCoords(pm);
+  const int ndim = v.GetNdim();
+  const Real dx = coords.Dx(X1DIR);
+  for (int i = X2DIR; i <= ndim; i++) {
+    const Real dy = coords.Dx(i);
+    PARTHENON_REQUIRE_THROWS(dx == dy,
+                             "UpdatePhi requires that DX be equal in all directions.");
+  }
+
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "UpdatePhi", DevExecSpace(), 0, v.GetDim(5) - 1, kb.s, kb.e,
+      jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        dv(b, 0, k, j, i) = -v(b, irho, k, j, i) * dx * dx;
+        dv(b, 0, k, j, i) += v(b, iphi, k, j, i - 1) + v(b, iphi, k, j, i + 1);
+        if (ndim > 1) {
+          dv(b, 0, k, j, i) += v(b, iphi, k, j - 1, i) + v(b, iphi, k, j + 1, i);
+          if (ndim == 3) {
+            dv(b, 0, k, j, i) += v(b, iphi, k - 1, j, i) + v(b, iphi, k - 1, j, i);
+          }
+        }
+        dv(b, 0, k, j, i) /= 2.0 * ndim;
+        dv(b, 0, k, j, i) -= v(b, iphi, k, j, i);
+      });
+
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "UpdatePhi", DevExecSpace(), 0, dv.GetDim(5) - 1, kb.s, kb.e,
+      jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        v(b, iphi, k, j, i) += dv(b, 0, k, j, i);
+      });
+
+  Kokkos::Profiling::popRegion(); // Task_Poisson_UpdatePhi
+  return TaskStatus::complete;
+}
+
+template <typename T>
+TaskStatus CheckConvergence(T *u, T *du) {
+  Kokkos::Profiling::pushRegion("Task_Poisson_UpdatePhi");
+  auto pm = u->GetParentPointer();
+
+  IndexRange ib = u->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = u->GetBoundsK(IndexDomain::interior);
+
+  std::vector<std::string> vars({"potential"});
+  auto v = u->PackVariables(vars);
+  auto dv = du->PackVariables(vars);
+
+  Real max_err;
+  parthenon::par_reduce(
+      parthenon::loop_pattern_mdrange_tag, "CheckConvergence", DevExecSpace(), 0,
+      v.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &eps) {
+        Real reps = std::abs(dv(b, 0, k, j, i) / v(b, 0, k, j, i));
+        Real aeps = std::abs(dv(b, 0, k, j, i));
+        eps = std::max(eps, std::min(reps, aeps));
+      },
+      Kokkos::Max<Real>(max_err));
+
```

Comment:
I understand you didn't think about MPI for this. But presumably a real application would need an MPI reduce here too? Maybe we should add a comment to that effect? And raise an error if the user tries to use MPI?

Reply:
But I _did_ think about MPI.  This all works in parallel.  The iteration completes only when _all_ completion_criteria tasks return `TaskStatus::complete`

Reply:
Ah I see. This makes sense.